### ISSUE TITLE
Pylance watches too many files, which explode the system inotify limit

### ISSUE THREAD

**Issue by ladyrick** [microsoft/pylance-release#5639]
Type: <b>Performance Issue</b>

# Pylance watches the entire /usr/local/lib/python3.10/dist-packages
Pylance watches too many files, which exceed the limit of system inotify limit.
However, my vscode works in docker environment and I cannot change the outside inotify limit.
So, basically my vscode's file watcher is broken.

I also tried to add dist-packages to my setting. still doesn't work.
```
"files.watcherExclude": {
  "**/.git/objects/**": true,
  "**/.git/subtree-cache/**": true,
  "**/.hg/store/**": true,
  "**/.vscode/**": true,
  "**/__pycache__/**": true,
  "**/node_modules/**": true,
  "**/vendor/**": true,
  "**/dist-packages/**": true,
}
```

Pylance watches all file in python search path to detect when user install new package. I think this is far from elegant, even kinda stupid.

Extension version: 2024.3.1
VS Code version: Code 1.87.2 (863d2581ecda6849923a2118d93a088b0745d9d6, 2024-03-08T15:20:57.520Z)
OS version: Darwin arm64 23.2.0
Modes:
Remote OS version: Linux x64 4.19.91-kangaroo.2.10.11.6c7cc891b.alios7.x86_64
Connection to 'ssh-remote+devbox2' could not be established

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Apple M1 Pro (10 x 24)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: enabled<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled|
|Load (avg)|4, 4, 4|
|Memory (System)|32.00GB (3.30GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|

|Item|Value|
|---|---|
|Remote|SSH: dsw|
|OS|Linux x64 4.19.91-kangaroo.2.10.11.6c7cc891b.alios7.x86_64|
|CPUs|Intel(R) Xeon(R) Processor (36 x 2600)|
|Memory (System)|465.02GB (458.89GB free)|
|VM|50%|

Connection to 'ssh-remote+devbox2' could not be established
</details>

**Comment by rchiodo** [comment]
We don't actually implement the file watchers, this is actually done through VS code. Your watcherExclude settings should be applied to VS code when it creates the watchers.

See this issue for getting your file watcher exclude globs to work:
https://github.com/microsoft/vscode/issues/175801

You might have to explicitly remove the full path to the `/usr/local/lib/python3.10/dist-packages`

See these other dupe issues on Pylance:
https://github.com/microsoft/pylance-release/issues/3001
https://github.com/microsoft/pylance-release/issues/2974

**Comment by chrispy-snps** [comment]
I'm glad I found this thread! I had the same issue.

For my VSCode development work, I use an SSH tunnel connection into RHEL7 Linux machines in my company's compute datacenter. When the Pylance extension was enabled, I would get messages like these:

```
[error] [File Watcher (parcel)] Inotify limit reached (ENOSPC) (path: /shared/python-3.9.2/lib/python3.9)
```

My company preinstalls many packages in the `site-packages/` directory of our Python installations:

```
$ find /shared/python-3.9.2/lib/python3.9/ | wc -l
109847
$ find /shared/python-3.9.2/lib/python3.9/site-packages/ | wc -l
102402
```

I tried to exclude the entire Python installation directory via an absolute path glob in `files.watcherExclude`, but that did not work.

However, thanks to the link that @rchiodo posted, I added an exclusion glob patterl *relative* to the directory in the watcher error message:

```
site-packages/**
```

and that worked!
